### ISSUE TITLE
Phase 1: Class Diagram and Data flow,

### ISSUE THREAD

**Issue by steveneliuk** [KhronosGroupArchives/webcl-validator#50]
Reviewing the codebase and documentation there doesnt seem to be a clear picture of the inner workings of the validator. There is indeed a picture painted, concerning the outcome, but not the process.

I believe it would be beneficial to have a class diagram and the flow of data between various classes detailed. Likewise, the dependencies on specific aspects of Clang and LLVM should clearly be marked. For instance, if a class inherits from the aforementioned, please specify it. If there are multiple parsing passes, specify the purpose, etc.

For the developer, this is clearly understood and shouldnt take much time to construct. These diagrams / details would provide the needed information to gain quick conceptual understanding of the validator and the processes that entail it.

This type of documentation should have been included during the completion of phase 1,

**Comment by elhigu** [comment]
Included to Phase IIa deliverables.

**Comment by elhigu** [comment]
This took a bit more time that I expected. However now I finally got it ready. 

This does not include class diagram, but describes clang classes that were used to implement certain passes (e.g. if they are matcher based or ast visitors). I also included examples about different steps of instrumentation trying to further explain lower level details of the algorithm. Also communication between passes and what data is stored in different places are shown. 

Diagram was done with draw.io webservice and also source xml is stored in doc/ for editing.

https://github.com/KhronosGroup/webcl-validator/blob/master/doc/InstrumentationFlowChart.pdf
